### ISSUE TITLE
resilver finish email too soon / shows replacing vdev

### ISSUE THREAD

**Issue by rlaager** [openzfs/zfs#9740]
### System information
Type | Version/Name
 --- | --- 
Distribution Name	| Ubuntu
Distribution Version	| 18.04
Linux Kernel	| 4.18.0-15-generic
Architecture	| amd64
ZFS Version	| 0.8.0\~rc4-1\~rlaager1.18.04.1 (a backport I made)
SPL Version	| 0.8.0\~rc4-1\~rlaager1.18.04.1

### Describe the problem you're observing

`zed` just sent me a "resilver finish" email, but the `zpool status` output in the email still shows the `replacing` vdev. A subsequent `zpool status` a few minutes later (as soon as I got to the system) shows it gone, as expected. Should there be some sleep on the "resilver finish" event, I wonder?

### Describe how to reproduce the problem

I suppose you'd have to start a resilver and wait for it to finish.

### Include any warning/errors/backtraces from the system logs

Here is the body of the email:
```
ZFS has finished a resilver:

   eid: 84578
 class: resilver_finish
  host: grape
  time: 2019-12-17 12:44:18-0600
  pool: grape1
 state: DEGRADED
status: Some supported features are not enabled on the pool. The pool can
	still be used, but some features are unavailable.
action: Enable all features using 'zpool upgrade'. Once this is done,
	the pool may no longer be accessible by software that does not support
	the features. See zpool-features(5) for details.
  scan: resilvered 1.57T in 0 days 13:54:50 with 0 errors on Tue Dec 17 12:44:18 2019
config:

	NAME             STATE     READ WRITE CKSUM
	grape1           DEGRADED     0     0     0
	  raidz3-0       DEGRADED     0     0     0
	    sdf1         ONLINE       0     0     0
	    sdc1         ONLINE       0     0     0
	    sdb1         ONLINE       0     0     0
	    sdd1         ONLINE       0     0     0
	    sda1         ONLINE       0     0     0
	    replacing-5  DEGRADED     0     0     0
	      old        FAULTED      9   220     0  too many errors
	      sdg1       ONLINE       0     0     0
	    sde1         ONLINE       0     0     0

errors: No known data errors
```

**Comment by behlendorf** [comment]
Duplicate of #9590.  It sounds like we're generating the finish event a little too early.  It looks like minimal fix for this would be to move the resilver finish event after the vdev replacement.  Something like the following (untested): 

```diff
diff --git a/module/zfs/dsl_scan.c b/module/zfs/dsl_scan.c
index 1c31261..4031089 100644
--- a/module/zfs/dsl_scan.c
+++ b/module/zfs/dsl_scan.c
@@ -903,9 +903,10 @@ dsl_scan_done(dsl_scan_t *scn, boolean_t complete, dmu_tx_t *
                        vdev_dtl_reassess(spa->spa_root_vdev, tx->tx_txg,
                            scn->scn_phys.scn_max_txg, B_TRUE);
 
-                       spa_event_notify(spa, NULL, NULL,
-                           scn->scn_phys.scn_min_txg ?
-                           ESC_ZFS_RESILVER_FINISH : ESC_ZFS_SCRUB_FINISH);
+                       if (scn->scn_phys.scn_min_txg == 0) {
+                               spa_event_notify(spa, NULL, NULL,
+                                   ESC_ZFS_SCRUB_FINISH);
+                       }
                } else {
                        vdev_dtl_reassess(spa->spa_root_vdev, tx->tx_txg,
                            0, B_TRUE);
diff --git a/module/zfs/spa.c b/module/zfs/spa.c
index 6cadefe..c77f494 100644
--- a/module/zfs/spa.c
+++ b/module/zfs/spa.c
@@ -7709,6 +7709,8 @@ spa_vdev_resilver_done(spa_t *spa)
        }
 
        spa_config_exit(spa, SCL_ALL, FTAG);
+
+       spa_event_notify(spa, NULL, NULL, ESC_ZFS_RESILVER_FINISH);
 }
 
 /*
```

@jwpoduska was just working in this area of the code and may have some thoughts about this.  We might want to do something a little more extensive and similarly move the scrub finish event.

**Comment by jwpoduska** [comment]
@behlendorf I don't think that would be quite correct. For example, currently, that event is only created if the 'complete' flag is set (and no pool checkpoint). Also, spa_vdev_resilver_done() gets called even when there isn't a resilver. I think that the function should really be called something like, spa_vdev_reassess_spares() (along with SPA_ASYNC_REASSESS_SPARES).

But, perhaps something like this might work:
```
diff --git a/module/zfs/spa.c b/module/zfs/spa.c
index 6cadefe..196bde7 100644
--- a/module/zfs/spa.c
+++ b/module/zfs/spa.c
@@ -7709,6 +7709,11 @@ spa_vdev_resilver_done(spa_t *spa)
        }

        spa_config_exit(spa, SCL_ALL, FTAG);
+
+       if (spa->spa_async_scan_notify) {
+               spa_event_notify(spa, NULL, NULL, spa->spa_async_scan_notify);
+               spa->spa_async_scan_notify = NULL;
+       }
```
(might need lock, but I doubt it would make too much difference)
```
diff --git a/module/zfs/dsl_scan.c b/module/zfs/dsl_scan.c
index 1c31261..60cc836 100644
--- a/module/zfs/dsl_scan.c
+++ b/module/zfs/dsl_scan.c
@@ -903,9 +903,8 @@ dsl_scan_done(dsl_scan_t *scn, boolean_t complete, dmu_tx_t *tx)
                        vdev_dtl_reassess(spa->spa_root_vdev, tx->tx_txg,
                            scn->scn_phys.scn_max_txg, B_TRUE);

-                       spa_event_notify(spa, NULL, NULL,
-                           scn->scn_phys.scn_min_txg ?
-                           ESC_ZFS_RESILVER_FINISH : ESC_ZFS_SCRUB_FINISH);
+                       spa->spa_async_scan_notify = scn->scn_phys.scn_min_txg ?
+                           ESC_ZFS_RESILVER_FINISH : ESC_ZFS_SCRUB_FINISH;
                } else {
                        vdev_dtl_reassess(spa->spa_root_vdev, tx->tx_txg,
                            0, B_TRUE);
```
```
@@ -298,6 +298,7 @@ struct spa {
        int             spa_async_suspended;    /* async tasks suspended */
        kcondvar_t      spa_async_cv;           /* wait for thread_exit() */
        uint16_t        spa_async_tasks;        /* async task mask */
+       const char      *spa_async_scan_notify;
        uint64_t        spa_missing_tvds;       /* unopenable tvds on load */
        uint64_t        spa_missing_tvds_allowed; /* allow loading spa? */
```

**Comment by behlendorf** [comment]
That's definitely better, although it appears it would still suffer from the issue of not posting and event when a checkpoint exists.  The rest makes logical sense, including the rename, but I'd have to test the various corner cases.
